solidity/contracts/token/extensions/HypERC4626.sol

```diff
@@ -71,32 +136,15 @@ contract HypERC4626 is HypERC20 {
         super._handle(_origin, _sender, _message);
     }
 
-    // Override to send shares locally instead of assets
-    function transfer(
-        address to,
+    /// override _transfer to handle share amounts internally but emit asset amounts
+    /// @notice This maintains internal share-based accounting while providing asset-based transfer events
+    /// @inheritdoc ERC20Upgradeable
+    function _transfer(
+        address sender,
+        address recipient,
         uint256 amount
-    ) public virtual override returns (bool) {
-        address owner = _msgSender();
-        _transfer(owner, to, assetsToShares(amount));
-        return true;
-    }
-
-    function shareBalanceOf(address account) public view returns (uint256) {
-        return super.balanceOf(account);
-    }
-
-    function balanceOf(
-        address account
-    ) public view virtual override returns (uint256) {
-        uint256 _balance = super.balanceOf(account);
-        return sharesToAssets(_balance);
-    }
-
-    function assetsToShares(uint256 _amount) public view returns (uint256) {
-        return _amount.mulDiv(PRECISION, exchangeRate);
-    }
-
-    function sharesToAssets(uint256 _shares) public view returns (uint256) {
-        return _shares.mulDiv(exchangeRate, PRECISION);
+    ) internal virtual override {
+        super._transfer(sender, recipient, amount);
```

Comment:
does this not already emit an event?

Reply:
It does but the event will shares amount as transferAmount in the event which is not desirable

Reply:
In `super._transfer()`, an event is still emitted showing the share amount. To avoid duplicate event calls, the `super._transfer()` call should be removed, and the internal logic needs to be implemented directly within the `_transfer()` override function.

Reply:
_balances is private in ERC20Upgradeable, so I cannot update it in the overriding function _transfer in HypERC4626, I can update the event signature to be TransferAssets() maybe?

Reply:
Yes, that is one approach, but the concern arises when handling the `Transfer` event in off-chain (e.g., on a CEX). They might call `transfer(from, to, 10)`, but the `Transfer()` event would emit the `share` value instead of 10. While it's unlikely that an ERC4626 token will be directly listed on a CEX, it would still be prudent to add a note about this in the documentation or function comments to prevent any unforeseen issues.

Another solution could be accessing `_balances` through the storage slot. https://programtheblockchain.com/posts/2018/03/09/understanding-ethereum-smart-contract-storage/

Reply:
Yes, that is one approach, but the concern arises when handling the `Transfer` event off-chain (e.g., on a CEX). They might call `transfer(from, to, 10)`, but the `Transfer()` event would emit the `share` value instead of 10. While it's unlikely that an ERC4626 token will be directly listed on a CEX, it would still be prudent to add a note about this in the documentation or function comments to prevent any unforeseen issues.

Another solution could be accessing `_balances` through the storage slot.
https://programtheblockchain.com/posts/2018/03/09/understanding-ethereum-smart-contract-storage/

Reply:
After talking with the team, we decided that we prefer to just omit the event and document in the contract that the amount in the event is the share amount, not the asset amount.